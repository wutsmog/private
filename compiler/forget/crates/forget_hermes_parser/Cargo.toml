[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }