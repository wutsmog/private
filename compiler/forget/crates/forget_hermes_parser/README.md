# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.