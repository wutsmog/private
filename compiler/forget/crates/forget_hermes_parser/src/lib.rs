mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
