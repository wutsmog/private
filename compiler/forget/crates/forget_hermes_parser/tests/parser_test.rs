use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
