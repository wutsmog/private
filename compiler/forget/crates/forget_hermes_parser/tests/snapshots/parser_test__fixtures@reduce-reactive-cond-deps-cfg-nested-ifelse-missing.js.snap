---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/reduce-reactive-cond-deps-cfg-nested-ifelse-missing.js
---
Input:
// props.a.b should NOT be added as a unconditional dependency to the reactive
// scope that produces x if it is not accessed in every path

function TestCondDepInNestedIfElse(props, other) {
  const x = {};
  if (foo(other)) {
    if (bar()) {
      x.a = props.a.b;
    }
  } else {
    x.d = props.a.b;
  }
  return x;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "TestCondDepInNestedIfElse",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 150,
          "end": 175
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 176,
            "end": 181
          }
        },
        {
          "type": "Identifier",
          "name": "other",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 183,
            "end": 188
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 200,
                    "end": 201
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 204,
                    "end": 206
                  }
                },
                "loc": null,
                "range": {
                  "start": 200,
                  "end": 206
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 194,
              "end": 207
            }
          },
          {
            "type": "IfStatement",
            "test": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "foo",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 214,
                  "end": 217
                }
              },
              "arguments": [
                {
                  "type": "Identifier",
                  "name": "other",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 218,
                    "end": 223
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 214,
                "end": 224
              }
            },
            "consequent": {
              "type": "BlockStatement",
              "body": [
                {
                  "type": "IfStatement",
                  "test": {
                    "type": "CallExpression",
                    "callee": {
                      "type": "Identifier",
                      "name": "bar",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 236,
                        "end": 239
                      }
                    },
                    "arguments": [],
                    "loc": null,
                    "range": {
                      "start": 236,
                      "end": 241
                    }
                  },
                  "consequent": {
                    "type": "BlockStatement",
                    "body": [
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "AssignmentExpression",
                          "operator": "=",
                          "left": {
                            "type": "MemberExpression",
                            "object": {
                              "type": "Identifier",
                              "name": "x",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 251,
                                "end": 252
                              }
                            },
                            "property": {
                              "type": "Identifier",
                              "name": "a",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 253,
                                "end": 254
                              }
                            },
                            "computed": false,
                            "loc": null,
                            "range": {
                              "start": 251,
                              "end": 254
                            }
                          },
                          "right": {
                            "type": "MemberExpression",
                            "object": {
                              "type": "MemberExpression",
                              "object": {
                                "type": "Identifier",
                                "name": "props",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 257,
                                  "end": 262
                                }
                              },
                              "property": {
                                "type": "Identifier",
                                "name": "a",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 263,
                                  "end": 264
                                }
                              },
                              "computed": false,
                              "loc": null,
                              "range": {
                                "start": 257,
                                "end": 264
                              }
                            },
                            "property": {
                              "type": "Identifier",
                              "name": "b",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 265,
                                "end": 266
                              }
                            },
                            "computed": false,
                            "loc": null,
                            "range": {
                              "start": 257,
                              "end": 266
                            }
                          },
                          "loc": null,
                          "range": {
                            "start": 251,
                            "end": 266
                          }
                        },
                        "directive": null,
                        "loc": null,
                        "range": {
                          "start": 251,
                          "end": 267
                        }
                      }
                    ],
                    "loc": null,
                    "range": {
                      "start": 243,
                      "end": 273
                    }
                  },
                  "alternate": null,
                  "loc": null,
                  "range": {
                    "start": 232,
                    "end": 273
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 226,
                "end": 277
              }
            },
            "alternate": {
              "type": "BlockStatement",
              "body": [
                {
                  "type": "ExpressionStatement",
                  "expression": {
                    "type": "AssignmentExpression",
                    "operator": "=",
                    "left": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 289,
                          "end": 290
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "d",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 291,
                          "end": 292
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 289,
                        "end": 292
                      }
                    },
                    "right": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "MemberExpression",
                        "object": {
                          "type": "Identifier",
                          "name": "props",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 295,
                            "end": 300
                          }
                        },
                        "property": {
                          "type": "Identifier",
                          "name": "a",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 301,
                            "end": 302
                          }
                        },
                        "computed": false,
                        "loc": null,
                        "range": {
                          "start": 295,
                          "end": 302
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "b",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 303,
                          "end": 304
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 295,
                        "end": 304
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 289,
                      "end": 304
                    }
                  },
                  "directive": null,
                  "loc": null,
                  "range": {
                    "start": 289,
                    "end": 305
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 283,
                "end": 309
              }
            },
            "loc": null,
            "range": {
              "start": 210,
              "end": 309
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "x",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 319,
                "end": 320
              }
            },
            "loc": null,
            "range": {
              "start": 312,
              "end": 321
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 190,
          "end": 323
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 141,
        "end": 323
      },
      "loc": null,
      "range": {
        "start": 141,
        "end": 323
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 141,
    "end": 323
  }
}
